Some fixes for common bugs in HttpClient

alxhub opened the pull request.

alxhub committed:
fix(common): accept falsy values as HTTP bodies

Previously, HttpClient used the overly clever test "body || null"
to determine when a body parameter was provided. This breaks when
the valid bodies '0' or 'false' are provided.

This change tests directly against 'undefined' to detect the presence
of the body parameter, and thus correctly allows falsy values through.

Fixes #19825.
Fixes #19195.

alxhub committed:
fix(common): remove useless guard in HttpClient

An invalid "if" condition is always true, and is thus useless. This
change removes it. No behavior changes.

Fixes #19223.

alxhub committed:
fix(common): treat an empty body as null when parsing JSON in HttpClient

Previously, XhrBackend would call JSON.parse('') if the response body was
empty (a 200 status code with content-length 0). This changes the XhrBackend
to attempt the JSON parse only if the response body is non-empty. Otherwise,
the body is left as null.

Fixes #18680.
Fixes #19413.
Fixes #19502.
Fixes #19555.

mary-poppins commented:
You can preview 62c7b98 at https://pr19958-62c7b98.ngbuilds.io/.

gkalpak commented on packages/common/http/test/response_spec.ts:
```diff
@@ -40,6 +40,11 @@ export function main() {
         expect(resp.url).toBeNull();
       });
+      it('accepts a falsy body', () => {
+        const resp = new HttpResponse({body: false});
```
Unused?

alxhub replied:
Removed.

gkalpak commented on packages/common/http/src/xhr.ts:
```diff
@@ -180,24 +180,20 @@ export class HttpXhrBackend implements HttpBackend {
-          } catch (error) {
-            // Cannot be certain that the body was meant to be parsed as JSON.
-            // Leave the body as a string.
+            // If this was an error request to begin with, leave it as a string, it probably
```
I wonder if it is worth preserving the `XSSI_PREFIX` in this case 🤔

alxhub replied:
Yes, done in a new commit.

gkalpak reviewed (commented):
The 4th commit also closes PR #19917 btw.

alxhub committed:
fix(common): don't strip XSSI prefix for if error isn't JSON

This changes XhrBackend to not strip the XSSI prefix from error text
if such a prefix is present but the remaining body does not parse as
JSON.

mary-poppins commented:
You can preview 45006a5 at https://pr19958-45006a5.ngbuilds.io/.

gkalpak reviewed (commented):
The 4th commit has incorrect scope (`service-worker` instead of `common`).

mary-poppins commented:
You can preview 4fa1792 at https://pr19958-4fa1792.ngbuilds.io/.

quanterion commented:
When is it going to be merged? Without those fixes it's impossible to use most HTTP DELETE requests that returns empty body

Maximaximum commented:
Hopefully this PR will get released as soon as possible, because the empty body parse issue is a real show stopper.

Taha-Di-Nero commented:
When this fix will be released ???
I spent to much time to migrate to the new HttpClient to downgrade.................

valkoun commented:
FYI, in the meantime you have 3 options to workaround:

1) Make sure responses always return a response body with 200 responses.
2) Make each response with no response body return a 204 No Content code.
3) Set `responseType: 'text'` for each http method call:
```
return this.http.post('api/groups', group, {
      responseType: 'text',
});
```

Maximaximum commented:
@valkoun yeah, thanks, but it's a whole lot of stupid error-prone work, so I hope this fix will get into release ASAP

Taha-Di-Nero commented:
Meanwhile this will be merged,  I fixed the problem on the server side returning http status 204 for all void service method and so no more parsing error.

realshadow commented:
Will this also get merged back to 4.4.6 or 7?

gigioSouza commented:
When will it be released?

vinayakpatil commented:
@alxhub Will this be merged anytime soon?

alxhub commented:
Yes, hopefully next week.

On Sat, Nov 25, 2017, 7:55 AM Vinayak Patil <notifications@github.com>
wrote:

> @alxhub <https://github.com/alxhub> Will this be merged anytime soon?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/19958#issuecomment-346923015>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD-1JYQglTdz-O3hJtlLgkfl3mY9hx6Uks5s57nzgaJpZM4QH_of>
> .
>

mary-poppins commented:
You can preview 4a4cb9f at https://pr19958-4a4cb9f.ngbuilds.io/.

hspier commented:
Any chance we can get a 4.4.7 release with this fix soon?
